src/app/clusters/basic/basic.cpp

```diff
@@ -13,48 +13,95 @@
  *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
- */
-
-/**
- *
- *    Copyright (c) 2020 Silicon Labs
- *
- *    Licensed under the Apache License, Version 2.0 (the "License");
- *    you may not use this file except in compliance with the License.
- *    You may obtain a copy of the License at
  *
- *        http://www.apache.org/licenses/LICENSE-2.0
- *
- *    Unless required by applicable law or agreed to in writing, software
- *    distributed under the License is distributed on an "AS IS" BASIS,
- *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- *    See the License for the specific language governing permissions and
- *    limitations under the License.
  */
-/****************************************************************************
- * @file
- * @brief Implementation for the Basic Server Cluster
- *plugin.
- *******************************************************************************
- ******************************************************************************/
+
 #include "basic.h"
 
-#include "af.h"
-#include <app/util/attribute-storage.h>
 #include <platform/CHIPDeviceLayer.h>
 
+#include <app/util/af.h>
+#include <app/util/attribute-storage.h>
+#include <gen/att-storage.h>
+#include <gen/attribute-id.h>
+#include <gen/attribute-type.h>
+#include <gen/cluster-id.h>
+#include <protocols/interaction_model/Constants.h>
+
+#include <cstring>
+
 using namespace chip;
+using namespace chip::DeviceLayer;
 
-bool emberAfBasicClusterResetToFactoryDefaultsCallback(void)
+namespace {
+template <size_t BufferLength>
+uint8_t * MakeZclCharString(uint8_t (&zclString)[BufferLength], char (&cString)[BufferLength])
 {
-    emberAfBasicClusterPrintln("RX: ResetToFactoryDefaultsCallback");
-    emberAfResetAttributes(emberAfCurrentEndpoint());
-    emberAfPluginBasicResetToFactoryDefaultsCallback(emberAfCurrentEndpoint());
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
-    return true;
+    static_assert(BufferLength <= 256, "Too long string to fit in ZCL_CHAR_STRING type");
+    zclString[0] = static_cast<uint8_t>(strlen(cString));
+    memcpy(&zclString[1], cString, zclString[0]);
+    return zclString;
 }
+} // namespace
 
-void emberAfPluginBasicResetToFactoryDefaultsCallback(EndpointId endpoint)
+void emberAfBasicClusterServerInitCallback(chip::EndpointId endpoint)
 {
-    DeviceLayer::ConfigurationMgr().InitiateFactoryReset();
+    uint16_t imVersion = chip::Protocols::InteractionModel::kVersion;
+    uint16_t vendorId;
+    uint16_t productId;
+    uint16_t productRevision;
+    uint32_t firmwareRevision;
+    char cString[65];
+    uint8_t zclString[65];
+
+    emberAfWriteAttribute(endpoint, ZCL_BASIC_CLUSTER_ID, ZCL_INTERACTION_MODEL_VERSION_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
```

Comment:
I guess this works for now if we need this ASAP, but shouldn't all this data actually be part of the Zap configuration and get spit out as default attribute values for singleton readonly attributes?  Then we can avoid duplicating this state per-endpoint, avoid storing it in RAM (because it would just live in .rodata), etc....

That means not storing it in the configuration manager, of course, but that seems ok... 

@vivien-apple thoughts?

Reply:
@bzbarsky-apple Currently, even read-only attributes are stored in RAM, although there is a separate variable for singleton attributes so that they are not duplicated per-endpoint. I've seen you created an issue to add support for "fixed" attributes which could be returned directly from the attributes' metadata.
My concern is that currently, the generated code for a single example is shared across all platforms for the example (e.g. `examples/lock-app/lock-common/gen`) so following your suggestion would require per-platform duplication of the code right? Unless we support setting the attribute defaults in the ZAP configuration to compile-time constants... That would also be nice since some Basic cluster attributes like the software version could be automatically generated with `git describe` etc. If we required people to update the ZAP file to change the software version it would force developers to put an extra effort in updating it manually.

Reply:
> My concern is that currently, the generated code for a single example is shared across all platforms for the example (e.g. `examples/lock-app/lock-common/gen`) so following your suggestion would require per-platform duplication of the code right? Unless we support setting the attribute defaults in the ZAP configuration to compile-time constants... That would also be nice since some Basic cluster attributes like the software version could be automatically generated with `git describe` etc. If we required people to update the ZAP file to change the software version it would force developers to put an extra effort in updating it manually.

To my knowledge the only thing we would need to duplicate is `endpoint_config.h`. Ideally a single source of trust, the `.zap` file should be used. But looking at our current setup, maybe some overriding mechanism of ZAP could be used.

For example, instead of duplicating the ZAP file to contains the correct value a mechanism to override the `default` field based on a some json definition can be used. Then the specific platform `endpoint_config.h` could be used.

@jepenven-silabs @tecimovic Any suggestions ?

Reply:
> Basic cluster attributes like the software version could be automatically generated with `git describe` etc.

There are a number of downsides to this:

- likely to trigger needless rebuilds or, it doesn't, leads to nondeterminism
- for the same reason, it breaks caching of build artifacts
- couples the build to "git"

Basically, it's just wrong from my perspective that updating documentation would cause firmware to rebuild. I am very sensitive to changes that have a large ripple effect on the build.

I wouldn't mind it as an option that's off by default for developers. I also wouldn't mind if we bumped a checked-in version stamp with some suitable frequency.

Reply:
@Damian-Nordic I agree that if we want to have a single set of generated stuff across all platforms then we might have to jump through some hoops if we have "fixed" data that differs on the different platforms.

That said, my understanding is that at some point we want to stop checking in the generated code and just generate at build time.  At that point, we could in fact have different things happening on different platforms pretty easily....

Reply:
> @Damian-Nordic I agree that if we want to have a single set of generated stuff across all platforms then we might have to jump through some hoops if we have "fixed" data that differs on the different platforms.
> 
> That said, my understanding is that at some point we want to stop checking in the generated code and just generate at build time. At that point, we could in fact have different things happening on different platforms pretty easily....

I'd suggest that Damien is on the right track here and ZAP is on the wrong one. We should be able to share an instance of a cluster between examples, or perhaps an entire endpoint, not centralize the entire device state in a global table. Cluster instances should bring their own storage (and should be able to be dynamically allocated) and be able to store read-only attributes in `constexpr` variables.

Does anyone even have an answer as to how we intend to make "bridges" or other dynamic devices work with this architecture?

Reply:
There are some general plans to change things around to allow dynamic devices.  It's clear that those can't really work with the current setup.